libraries/AP_ADSB/AP_ADSB_Sagetech.cpp

```diff
@@ -0,0 +1,539 @@
+#define SAGETECH_VALIDFLAG_VELOCITY         (1U<<2)
+#define SAGETECH_VALIDFLAG_GND_SPEED        (1U<<3)
+#define SAGETECH_VALIDFLAG_HEADING          (1U<<4)
+#define SAGETECH_VALIDFLAG_V_RATE_GEO       (1U<<5)
+#define SAGETECH_VALIDFLAG_V_RATE_BARO      (1U<<6)
+#define SAGETECH_VALIDFLAG_EST_LATLNG       (1U<<7)
+#define SAGETECH_VALIDFLAG_EST_VELOCITY     (1U<<8)
+
+// detect if any port is configured as Sagetech
+bool AP_ADSB_Sagetech::detect()
+{
+    return (AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_Sagetech_ADSB, 0) != nullptr);
+}
+
+bool AP_ADSB_Sagetech::init()
+{
+    if (_port == nullptr) {
+        _port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_Sagetech_ADSB, 0);
+    }
+
+    return (_port != nullptr);
+}
+
+void AP_ADSB_Sagetech::update()
+{
+    if (_port == nullptr) {
+        return;
+    }
+
+    const uint32_t now_ms = AP_HAL::millis();
+
+    // -----------------------------
+    // read any available data on serial port
+    // -----------------------------
+    uint32_t nbytes = MIN(_port->available(), 10 * PAYLOAD_XP_MAX_SIZE);
+    while (nbytes-- > 0) {
+        const uint8_t data = (uint8_t)_port->read();
+        if (parse_byte_XP(data)) {
+            handle_packet_XP(message_in_xp.packet);
+        }
+    } // while nbytes
+
+
+    // -----------------------------
+    // handle timers for generating data
+    // -----------------------------
+    if (!last_packet_initialize_ms || (now_ms - last_packet_initialize_ms >= 5000)) {
+        last_packet_initialize_ms = now_ms;
+        send_packet(MsgType_XP::Installation_Set);
+
+
+    } else if (!last_packet_PreFlight_ms || (now_ms - last_packet_PreFlight_ms >= 8200)) {
+        // send once, for now..
+        last_packet_PreFlight_ms = now_ms;
+        // TODO: allow callsign to not require a reboot
+        send_packet(MsgType_XP::Preflight_Set);
+
+    } else if (now_ms - last_packet_Operating_ms >= 1000 && (
+            last_packet_Operating_ms == 0 || // send once at boot
+            // send as data changes
+            last_operating_squawk != _frontend.out_state.cfg.squawk_octal ||
+            abs(last_operating_alt - _frontend._my_loc.alt) > 1555 ||      // 1493cm == 49ft. The output resolution is 100ft per bit
+            last_operating_rf_select != _frontend.out_state.cfg.rfSelect))
+    {
+        last_packet_Operating_ms = now_ms;
+        last_operating_squawk = _frontend.out_state.cfg.squawk_octal;
+        last_operating_alt = _frontend._my_loc.alt;
+        last_operating_rf_select = _frontend.out_state.cfg.rfSelect;
+        send_packet(MsgType_XP::Operating_Set);
+
+    } else if (now_ms - last_packet_GPS_ms >= (_frontend.out_state.is_flying ? 200 : 1000)) {
+        // 1Hz when not flying, 5Hz when flying
+        last_packet_GPS_ms = now_ms;
+        send_packet(MsgType_XP::GPS_Set);
+    }
+}
+
+void AP_ADSB_Sagetech::send_packet(const MsgType_XP type)

```

Comment:
These calls are only made in one place we can -  do without the indirection of send_packet

(this is probably because you removed the handling of ACKs).

Reply:
I'd like to keep this in because I may want to add queuing